src/ceph-volume/ceph_volume/util/metadata.py

```diff
@@ -0,0 +1,438 @@
+            dev_names (List[str]): List of devices that are on the host
+        """
+        if not worker_pool_size:
+            self.worker_pool_size = os.cpu_count()
+        else:
+            assert worker_pool_size <= os.cpu_count()
+            self.worker_pool_size = worker_pool_size
+
+        logger.info(f"MetadataCache using {self.worker_pool_size} threads")
+
+    def get_device_data(self) -> Dict[str, Any]:
+        """Use lsblk to discover devices on the host
+
+        The attributes of the devices will be used in the disk_api consumed by the
+        Device class
+
+        Raises:
+            OSError: lsblk problems, unable to continue
+            ValueError: Returned data looks to be in an incompatible format
+
+        Returns:
+            Dict[str, Any]: Dict of devices
+        """
+        field_list = set()
+        field_list.update(self.blk_field_list)
+        field_list.update(self.disk_field_list)
+
+         # using the J option to output the blk device hierarchy in json
+        lsblk_cmd = ['lsblk', '-bJo', ','.join(field_list)]
+        devices = {}
+        raw_out, err, rc = call(lsblk_cmd)
+        if rc > 0:
+            # lsblk cmd failed, can't continue, raise since this is fatal
+            raise OSError("OS call to lsblk returned {rc}")
+
+        # call returns the output having passed through splitlines, so we need to re-assemble
+        device_json = _js_reformat(raw_out)
+
+        if "blockdevices" not in device_json:
+            # can't continue
+            raise ValueError("blockdevices section missing from lsblk")
+
+        for dev_data in device_json.get("blockdevices"):
+
+            dev_name = dev_data.get('name')
+            if dev_name.startswith(self.excluded_devs):
+                continue
+
+            devices[dev_name] = dev_data
+        return devices
+
+    def get_lsm_devices(self, dev_name_list: List[str]) -> List[str]:
+        """Return a list of device paths that will be compatible with libstoragemgmt
+
+        Args:
+            dev_name_list (List[str]): list of device names e.g ['sda','sdb'...]
+
+        Returns:
+            List[str]: List of device paths that will work with lsm
+        """
+        return [f"/dev/{dev_name}" for dev_name in dev_name_list if dev_name.startswith(self.lsm_compatible_devices)]
+
+    def get_device_holders(self, dev_names: List[str]) -> Dict[str, List[str]]:
+        """Create a map of devices to their holders (children)
+
+        Args:
+            dev_names (List[str]): device name list
+
+        Returns:
+            Dict[str, List[str]]: device -> List of child devices (holders)
+        """
+        return {dev_name: get_holders(dev_name) for dev_name in dev_names}
+
+    def start_workers(self):
+        """Create a pool of worker threads
+        """
+        for t in range(self.worker_pool_size):
+            t = threading.Thread(target=worker)
+            t.setDaemon = True
+            t.start()
```

Comment:
I was going to suggest you the new Python 3 `asyncio` package, but since the underlying code is still blocking, I think we still need threading, so what about the Python [`concurrent.futures`](https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example) package? It'll save from using threads/queues/etc.:

```py
with concurrent.futures.ThreadPoolExecutor(max_workers=worker_pool_size) as executor:
  futures = {
      executor.submit(udevadm_property, dev, DEVID_PROPS): f'udev.{dev}' for
      dev in dev_paths}
  futures.update({
      executor.submit(blkid, dev): f'blkid.{dev}' for dev in dev_paths})
  ...
  # gather data
  output_data = {}
  for future in concurrent.futures.as_completed(futures):
    key = futures[future]
    try:
        output_data[key] = future.result()
    except Exception as e:
        print('%r generated an exception: %s' % (key, e))
    else:
        print('%r page is %d bytes' % (key, len(output_data[key])))
```

Reply:
Yes - I looked at concurrent.futures. But when I tested it, I found it was slower than the current lower-level implementation of a producer/consume/queue model - which goes against the primary goal which was to be as fast as possible.

Reply:
BTW - I also tried asyncio...same thing. Handcrafting the data gathering was quicker.

Reply:
Interesting finding, good to know that (I'm curious: how much slower?)... A good thing with the `concurrent.futures.Executor` was that you could easily switch from the GIL-aware `threading` to full-throttle real parallelization (`multiprocessing`) with the `ProcessPoolExecutor`.

And regarding `asyncio` the main downside is that you'd have to rewrite your code with the `async*` counterparts (e.g.: [`asyncio Subprocess`](https://docs.python.org/3/library/asyncio-subprocess.html)).